### ISSUE TITLE
bacteria contig

### ISSUE THREAD

**Issue by AlisaGU** [lastz/lastz#35]
Hi,
I have several bacteria genomes. There are several sequences in a single genome. Genome size is small, like 3M.
My aim is to get a multiple genome alignment. When using lastz to do pairwise alignment, I have some questions.
1. is it necessary to do chaining and netting? Due to a small size, I can align them directly and get .maf file. The code is like that
```
lastz genomes_repeatMasked/ps128.fna.masked[multiple] genomes_repeatMasked/299v.fna.masked[multiple] --ambiguous=iupac  --seed=12of19 --notransition H=2000 --format=maf> test.maf
```
2. README just says: `This action is required when the target (not the query) is comprised of multiple sequences.` I don't know if the code in the question 1 is right.

Could you give some tips?

Best regards,

**Comment by AlisaGU** [comment]
Is it correct if I concatenate contigs  into a sequence? Then, use this sequence as the input of lastz

**Comment by rsharris** [comment]
The command line you show in question 1 is OK. The note you quote is question 2 is meant to explain that the [multiple] action is discouraged for the query (the second filename is the query).

The chaining and netting was written by the folks at UCSC. I don't often do multiple alignment in my own work, so I'm not familiar with the details of those processes.

Lastz does have an option called chaining, but this is something different (but similar). Usually it is only appropriate if your sequences don't have any rearrangements.

As for concatenating contigs into a sequence, usually you should have a fasta file for each genome. If a genome has several contigs, each contig should have a header line in the fasta file, followed by its nucleotide sequence.

**Comment by AlisaGU** [comment]
1. I can use the "multiple" tag in query though it's discouraged.
2. chaining and netting is unnecessary for small genomes.
3. It's unnecessary to concatenate contigs into a sequence though it's not wrong.

Is it right?😀

**Comment by rsharris** [comment]
> 1. I can use the "multiple" tag in query though it's discouraged.

It's not optimal, and there's no reason to use it except in very rare cases. It will work, but there might be minor differences in alignment output.

> 2. chaining and netting is unnecessary for small genomes.

No idea. The chaining and netting steps aren't part of the lastz package. Whoever at UCSC is responsible for those would be the person to ask.

> 3. It's unnecessary to concatenate contigs into a sequence though it's not wrong.

You will make your life harder if you concatenate them. (1) You would have to post-process the alignments to map coordinates back to the original contigs. (2) You would have to do something to avoid chimeric segments in alignments.

Use one fasta file for each genome. If a genome has several contigs, each contig should begin with its own header line, followed by its nucleotide sequence. Header lines start with a >. The nucleotide sequence can be all one one line (but NOT on the line with the header) or spread over multiple lines.

**Comment by AlisaGU** [comment]
concatenating makes problem harder, but "multiple" is not suitable for query.
What should I do if the query genome has many contigs?
loop? Map one contig to the reference genome in one loop?

**Comment by rsharris** [comment]
http://www.bx.psu.edu/~rsharris/lastz/README.lastz-1.04.03.html#action_multiple

Lastz needs to have the entire target file in memory. For historical reasons and an engineering choice I made circa 2007, you have to use [multiple] to tell it the file contains more than one sequence and to load them all into memory.

Lastz processes the sequences in the query file one at a time. It loads only one into memory at a time, aligns it to the entire target, then loads the next sequence. It does the looping for you. If you use [multiple] with the query file it will load all the sequences into memory and treat it as one large query. Everything will still work correctly, but there is no reason to do this. There will be differences in the order in which alignments are reported, and when combined with some other options you will see minor differences in the alignments that are output.

To recap, the [multiple] action is discouraged for the query because it is not necessary for the query.

**Comment by AlisaGU** [comment]
I get it! The right code should be `lastz genomes_repeatMasked/ps128.fna.masked[multiple] genomes_repeatMasked/299v.fna.masked --ambiguous=iupac  --seed=12of19 --notransition H=2000 --format=maf> test.maf`.

I am grateful for your quick and kind answer. 
It is very helpful information for me.
